k2/csrc/aux_labels.cc

```diff
@@ -0,0 +1,66 @@
+// k2/csrc/aux_labels.cc
+
+// Copyright (c)  2020  Xiaomi Corporation (author: Haowen Qiu)
+
+// See ../../LICENSE for clarification regarding multiple authors
+
+#include "k2/csrc/aux_labels.h"
+
+#include <numeric>
+#include <vector>
+
+#include "glog/logging.h"
+#include "k2/csrc/fsa.h"
+
+namespace k2 {
+
+void MapAuxLabels1(const AuxLabels &labels_in,
+                   const std::vector<int32_t> &arc_map, AuxLabels *labels_out) {
+  CHECK_NOTNULL(labels_out);
+  auto &start_pos = labels_out->start_pos;
+  auto &labels = labels_out->labels;
+  start_pos.clear();
+  labels.clear();
+
+  int32_t num_labels = 0;
+  for (const auto &arc_index : arc_map) {
+    start_pos.push_back(num_labels);
+    int32_t pos_start = labels_in.start_pos[arc_index];
+    int32_t pos_end = labels_in.start_pos[arc_index + 1];
+    for (int32_t pos = pos_start; pos != pos_end; ++pos) {
+      int32_t label = labels_in.labels[pos];
+      DCHECK_NE(label, kEpsilon);
+      labels.push_back(label);
+      ++num_labels;
+    }
+  }
+  start_pos.push_back(num_labels);
+}
+
+void MapAuxLabels2(const AuxLabels &labels_in,
+                   const std::vector<std::vector<int32_t>> &arc_map,
+                   AuxLabels *labels_out) {
+  CHECK_NOTNULL(labels_out);
+  auto &start_pos = labels_out->start_pos;
+  auto &labels = labels_out->labels;
+  start_pos.clear();
+  labels.clear();
```

Comment:
Please pre-allocate space for `start_pos` and `labels`.

Reply:
We cannot actually pre-allocate here if we have not iterated over `arc_map`, as we don't know how many arcs in it.

Reply:
The maximum number of arcs is `labels_in.labels.size()`.
We can resize it at the end if extra memory is allocated.

Reply:
actually it's not (think about for log-sum version for most of algortihms, there may be many duplicate arcs in arc_derivs_out)